### PR TITLE
Threshold otpimization

### PR DESCRIPTION
**Pull request by Tilana**
This PR contains the following changes:
- the default threshold is set to 0.5. Therefore, even without treshold optimization the classifier results are used
- threshold optimization becomes is more flexible now:
  - f1 score is optimized based on a minimum precision level
  - if the best f1 scores is reached at different thresholds the least extreme / most central one is chosen. This makes threshold optimization more robust to variances in the dev set used for threshold optimization
  - topic info stores for each threshold the corresponding f1 score, precision and recall measures. This makes queries easier and makes separately storing the _recalls, thresholds_ as well as _precision-at-suggested, f1--at-suggested_ unnecessary
- quality scores are based on f1, same is true for confidence level for the prediction of a paragraph.
- a new test model is introduced. It has only 4 categories and is therefore easier to use for testing

### CODE REVIEW

File: app/TopicInfo.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+import json
```
Comment (samschaevitz) [inline app/TopicInfo.py]: nit: please name this file topic_info.py

File: tests/test_e2e.py [hunk 2]
```diff
@@ -14,40 +14,39 @@ def wait_for_task(client: FlaskClient, name: str) -> None:
...
 
 
 def test_e2e(app: Flask, fs: FakeFilesystem) -> None:
-    seq_pattern = (
-        'react more swiftly to comply with international instruments %d')
+    seq_pattern = ('Continue working for the eradiction of poverty %d')
     instance_path = './testdata/test_model/test_instance'
     fs.add_real_directory(instance_path)
-    fs.remove_object('./testdata/test_model/test_instance/quality.json')
+    # fs.remove_object('./testdata/test_model/test_instance/quality.json')
```
Comment (samschaevitz) [inline tests/test_e2e.py]: remove comment?

File: tests/test_routes.py [hunk 3]
```diff
@@ -93,13 +93,13 @@ def test_model_status(app: Flask, fs: FakeFilesystem) -> None:
...
-        'name': 'Asylum-seekers - refugees',
-        'quality': 1.0,
-        'samples': 260
-    }
-    assert result['topics']['National Human Rights Institution'] == {
-        'name': 'National Human Rights Institution',
-        'quality': 1.0,
-        'samples': 552
-    }
+    # assert result['topics']['Asylum-seekers - refugees'] == {
```
Comment (samschaevitz) [inline tests/test_routes.py]: remove (or change topic values?)

File: tests/test_threshold_optimization.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
+import numpy as np
+import pandas as pd
+import pytest
+
+from app.threshold_optimization import (ComputeThresholds, build_score_matrix,
+                                        compute_scores, optimize_threshold)
+
+
+class Test_thresholdOptimization:
```
Comment (samschaevitz) [inline tests/test_threshold_optimization.py]: nit: no mixing camelcase and underscores please e.g. TestThresholdOptimization

File: tests/test_threshold_optimization.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
+import numpy as np
+import pandas as pd
+import pytest
+
+from app.threshold_optimization import (ComputeThresholds, build_score_matrix,
```
Comment (samschaevitz) [inline tests/test_threshold_optimization.py]: nit: rename threshold_optimization to thresholds 
and
(optional) remove 'thresholds' from function names e.g. compute(), optimize()

**Comment by Tilana** [comment]
thanks, Sam, for your comments! I just updated the code to ensure a consistent naming scheme

File: tests/test_routes.py (hunk 3 above)
Comment (Tilana) [reply tests/test_routes.py]: I added the missing test in the latest commit

File: app/model_status.py [hunk 6]
```diff
@@ -86,20 +87,14 @@ def _build_status_dict(self) -> Dict[str, Any]:
...
-                            app.config['DESIRED_CLASSIFIER_PRECISION'], 0.0),
+                    'name': t,
+                    'samples': ti.num_samples,
+                    'quality': ti.get_quality(),
                 }
         return {
             'name': self.model_name,
             'instances': instances,
             'preferred': preferred,
-            'completeness': quality_at_precision.get('completeness', 0.0),
```
Comment (bdittes) [inline app/model_status.py]: Please keep these - they are the only way I personally understand how to judge the overall quality of the model...

File: app/topic_info.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        self.scores.index = self.scores.index.map(float)
+
+    def get_quality(self) -> Any:
+        return self.scores.loc[self.suggested_threshold].f1
+
+    def get_confidence_at_probability(self, prob: float) -> Any:
+        # TODO: include distance between threshold and probability
+        scores = self.scores[self.scores.index >= prob]
+        if len(scores) > 0:
+            quality = scores.iloc[0].f1
```
Comment (bdittes) [inline app/topic_info.py]: This is weird - by including the recall at the threshold (which f1 does), you're saying confidence_at_probability(1.0) is going to be much lower than 1?

File: tests/test_classifier.py [hunk 8]
```diff
@@ -14,19 +14,19 @@ def test_classify(self, fs: FakeFilesystem) -> None:
...
 
         assert c.instance == 'test_instance'
-        print(result)
         assert result
         # result ~ [{topic: probability, topic2: probability, ...}, ...]
         for topic, _ in result[0].items():
             assert topic in c.vocab
-        assert result[0]['Right to education'] >= 0.7
+        assert len(result) == 1
+        assert result[0]['Right to health'] >= 0.5
```
Comment (bdittes) [inline tests/test_classifier.py]: why such a low threshold? this should be a high-confidence sequence, right?

File: tests/test_e2e.py [hunk 9]
```diff
@@ -14,40 +14,38 @@ def wait_for_task(client: FlaskClient, name: str) -> None:
...
-        'react more swiftly to comply with international instruments %d')
+    seq_pattern = ('Continue working for the eradiction of poverty %d')
     instance_path = './testdata/test_model/test_instance'
     fs.add_real_directory(instance_path)
-    fs.remove_object('./testdata/test_model/test_instance/quality.json')
     fs.remove_object('./testdata/test_model/test_instance/thresholds.json')
     client = app.test_client()
     with app.test_request_context():
-        # initial classify returns no topics because we deleted thresholds.json.
+        # without threshold file default quality is set to 0.5
```
Comment (bdittes) [inline tests/test_e2e.py]: for all topics?

File: tests/test_thresholds.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+    def test_optimize(self) -> None:
+        scores = pd.DataFrame([(0.3, 0.2, 0.4), (0.7, 0.8, 0.6),
+                               (0.8, 0.7, 0.9)],
+                              columns=['f1', 'precision', 'recall'],
+                              index=[0.2, 0.4, 0.7])
+        assert optimize(scores) == 0.7
+        assert optimize(scores, min_prec=0.8) == 0.4
+        # if minimum precision cannot be achieved return default threshold
+        assert optimize(scores, min_prec=1.1) == 0.5
+
```
Comment (bdittes) [inline tests/test_thresholds.py]: Can you add some confidence_at_probability tests, please?

File: tests/test_e2e.py (hunk 9 above)
Comment (Tilana) [reply tests/test_e2e.py]: It's difficult to determine the quality of a prediction if we don't have data to optimize thresholds and evaluate performance. I can be super good, but also quite bad - so I hoped that with a default confidence 0.5 we could display this arbitrary decision...
Still I do see a point that for end users 0.5 might seem quite as a quite confident decision. I lowered it down to 0.3

File: tests/test_thresholds.py (hunk 10 above)
Comment (Tilana) [reply tests/test_thresholds.py]: Good point!
In the latest commit I just added tests for the TopicInfo class. Confidence_at_probability tests is also included.

File: tests/test_classifier.py (hunk 8 above)
Comment (Tilana) [reply tests/test_classifier.py]: the test still passes with an increased threshold. I just adapted that.

**Pull request merged**